src/OpenLoco/src/Map/MapGenerator/MapGenerator.cpp

```diff
@@ -433,13 +435,108 @@ namespace OpenLoco::World::MapGenerator
         }
     }
 
+    static void updateProgress(uint8_t value);
+
+    // 0x004595B7
+    static bool isCargoProducedAnywhere(uint8_t requiredCargoType)
+    {
+        auto commonBuildingCargoType = IndustryManager::getMostCommonBuildingCargoType();
+        if (commonBuildingCargoType == requiredCargoType)
+            return true;
+
+        for (auto i = 0U; i < ObjectManager::getMaxObjects(ObjectType::industry); i++)
+        {
+            auto* industryObj = ObjectManager::get<IndustryObject>(i);
+            if (industryObj == nullptr)
+            {
+                continue;
+            }
+
+            if (getCurrentYear() < industryObj->designedYear || getCurrentYear() > industryObj->obsoleteYear)
+            {
+                continue;
+            }
+
+            for (auto producedCargoType : industryObj->producedCargoType)
+            {
+                if (producedCargoType == requiredCargoType)
+                    return true;
+            }
+        }
+
+        return false;
+    }
+
     // 0x004597FD
     static void generateIndustries(uint32_t minProgress, uint32_t maxProgress)
     {
-        registers regs;
-        regs.eax = minProgress;
-        regs.ebx = maxProgress;
-        call(0x004597FD, regs);
+        auto numIndustriesAvailable = 0;
+        for (auto indObjId = 0U; indObjId < ObjectManager::getMaxObjects(ObjectType::industry); indObjId++)
+        {
+            if (ObjectManager::get<IndustryObject>(indObjId) != nullptr)
+            {
+                numIndustriesAvailable++;
+            }
+        }
+
+        if (!numIndustriesAvailable)
+            return;
+
+        CompanyManager::setUpdatingCompanyId(CompanyId::neutral);
+
+        auto progressTicksPerIndustry = (maxProgress - minProgress) / numIndustriesAvailable;
+        auto currentProgress = minProgress;
+
+        for (auto indObjId = 0U; indObjId < ObjectManager::getMaxObjects(ObjectType::industry); indObjId++)
+        {
+            auto* industryObj = ObjectManager::get<IndustryObject>(indObjId);
+            if (industryObj == nullptr)
+            {
+                continue;
+            }
+
+            currentProgress += progressTicksPerIndustry;
+            updateProgress(currentProgress);
+
+            // Check if industry is available at present
+            if (getCurrentYear() < industryObj->designedYear || getCurrentYear() > industryObj->obsoleteYear)
```

Comment:
is this really `>` in the assembly? it makes more sense for it to be `>=`

Reply:
You're right:
```asm
mov     cx, current_year
cmp     cx, [ebp+0CAh]  ; designedYear
jb      loc_4599A3      ; continue
cmp     cx, [ebp+0CCh]  ; obsoleteYear
jnb     loc_4599A3      ; continue
```
It's jnb (jump not below), so should be `>=`.